File: custom-sort-string.js
// Time Complexity : O(M + N) where M and N are the lengths of S and T
// Space Complexity : O(T)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {string} S
 * @param {string} T
 * @return {string}
 */
var customSortString = function(S, T) {
    if (!S || !S.length) return T;
    if (!T || !T.length) return '';
    
    let letterCount = new Array(26).fill(0);
    
    for (const [index, char] of [...T].entries()) {
        letterCount[T.charCodeAt(index) - 97]++;
    }
    let answer = '';
    
    for (const [index, char] of [...S].entries()) {
        let charCode = S.charCodeAt(index) - 97;
        for (let i = 0; i < letterCount[charCode]; i++) {
            answer += char;
        }
        letterCount[S.charCodeAt(index) - 97] = 0;
    }
    for (let i = 0; i <= 26; i++) {
        let currLetter = String.fromCharCode(i + 97);
        for (let j = 0; j < letterCount[i]; j++) {
            answer += currLetter;
        }
    }
    
    return answer;
};

File: customsortstring.swift
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
    func customSortString(_ order: String, _ s: String) -> String {
        if order == nil || order.count == 0 || s.count == 0 || s == nil {
            return "" 
        }
        var orderArray = Array(order)
        var sArray = Array(s)
        var output = ""
        var hashmap = [Character:Int]()
            for i in 0..<sArray.count {
                var char = sArray[i]
                hashmap[char,default: 1] += 1   
            }
        for i in 0..<orderArray.count {
            var char = orderArray[i]
            if let count = hashmap[char] {
                hashmap[char] = 0
                if count > 0 {
                    for j in 1..<count {
                        output.append(char)
                    }
                }
            }
        }
        for i in 0..<sArray.count {
            var char = sArray[i]
            if let count = hashmap[char] {
                if count > 0 {
                        output.append(char)
 
                }
            }
        }
        return output
    }